File: Coding_1215 1D_DP/최원준/최원준/4번.py
# https://leetcode.com/problems/palindromic-substrings/

'''
1. 아이디어 :
    i번째 캐릭터부터 양쪽으로 뻗어나간다.
    홀짝수 palindrome 둘다 체크한다.
2. 시간복잡도 :
    O(n^2)
3. 자료구조 :
    배열
'''

class Solution:
    def countSubstrings(self, s: str) -> int:

        def count_pal(idx,left,right,s):
            count=0
            while left>=0 and right<len(s) and s[left]==s[right]:
                count+=1
                left-=1
                right+=1
            return count

        ans = 0
        for i in range(len(s)):
            ans+=count_pal(i,i,i,s)
            ans+=count_pal(i,i,i+1,s)

        return ans

File: Coding_1215 1D_DP/최원준/이강훈/4번.java
/*
https://leetcode.com/problems/palindromic-substrings/submissions/

아이디어
3번 팰린드롬 문제의 간단 버전
 -> DP 저장, 팰린드롬 조건 일치 시 정답 수 증가

자료구조
2차원 배열

시간복잡도
O (N^2)
*/

public class Solution {
    public int countSubstrings(String s) {
        boolean[][] dy = new boolean[s.length()][s.length()];

        int maxLength = Integer.MIN_VALUE;
        int answer = 0;

        for (int i = 0; i < s.length(); i++) {
            dy[i][i] = true;
            answer++;

            for (int j = 0; j < i; j++) {
                if (s.charAt(i) == s.charAt(j) && (i - j < 2 || dy[j + 1][i - 1])) {
                    dy[j][i] = true;
                    answer++;
                }
            }
        }

        return answer;
    }
}
